### PR TITLE
fix: [PL-31963]: Added validation for inputs in custom secret manager

### PR DESCRIPTION
**Pull request by VikasMaddukuriHarness**
## Describe your changes
we are allowing Custom SM’s get created even though the runTime params are not being passed when creating SM (this can be done via yaml in UI or via api call). This should not be allowed and this is to be handled

Testing:
1. Tested the flow without passing the necessary run time params and verified that it is not allowing us to create the SM.
2. Tested the flow with passing all the run time params and verified it is getting created.
3. Checked the final params that are getting picked up when actually using the custom SM and verified they are taking values from template or not.
## Checklist
- [x] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/46878)

### CODE REVIEW

File: 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java [hunk 1]
```diff
@@ -155,6 +177,51 @@ private ConnectorResponseDTO createSecretManagerConnector(
     return defaultConnectorService.create(connector, accountIdentifier, NONE);
   }
 
+  private void testCustomSecretManagerInputs(ConnectorConfigDTO connectorConfigDTO, String accountIdentifier,
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java:180]: rename to validateCustomSecretManagerInputs

File: 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java [hunk 2]
```diff
@@ -147,6 +161,14 @@ private ConnectorResponseDTO createSecretManagerConnector(
     // Check which type of token is provided
     ngVaultService.processTokenLookup(connector, accountIdentifier);
 
+    // Validate inputs for custom secret manager
+    try {
+      testCustomSecretManagerInputs(connectorConfigDTO, accountIdentifier, connectorInfo.getOrgIdentifier(),
+          connectorInfo.getProjectIdentifier());
+    } catch (IOException ex) {
+      log.error("error reading template yaml");
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java:169]: Add more context to log

File: 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java [hunk 3]
```diff
@@ -155,6 +177,51 @@ private ConnectorResponseDTO createSecretManagerConnector(
     return defaultConnectorService.create(connector, accountIdentifier, NONE);
   }
 
+  private void testCustomSecretManagerInputs(ConnectorConfigDTO connectorConfigDTO, String accountIdentifier,
+      String orgIdentifier, String projectIdentifier) throws IOException {
+    if (connectorConfigDTO instanceof CustomSecretManagerConnectorDTO) {
+      // get the template params
+      String templateRef = ((CustomSecretManagerConnectorDTO) connectorConfigDTO).getTemplate().getTemplateRef();
+      //      Map<String, List<NameValuePairWithDefault>> inputValues = ((CustomSecretManagerConnectorDTO)
+      //      connectorConfigDTO).getTemplate().getTemplateInputs();
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java:186]: Remove commented code

File: 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java [hunk 4]
```diff
@@ -155,6 +177,51 @@ private ConnectorResponseDTO createSecretManagerConnector(
...
+      List<NameValuePairWithDefault> envVariables = ((CustomSecretManagerConnectorDTO) connectorConfigDTO)
+                                                        .getTemplate()
+                                                        .getTemplateInputs()
+                                                        .get("environmentVariables");
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java:190]: move `environmentVariables` to constant

File: 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java [hunk 5]
```diff
@@ -155,6 +177,51 @@ private ConnectorResponseDTO createSecretManagerConnector(
...
+      Set<String> inputs = new HashSet<>();
+      for (NameValuePairWithDefault node : envVariables) {
+        if (inputs.contains(node.getName())) {
+          log.error("Duplicate input.");
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java:194]: Add more context to log

File: 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java [hunk 6]
```diff
@@ -155,6 +177,51 @@ private ConnectorResponseDTO createSecretManagerConnector(
...
+          throw new InvalidRequestException("Duplicate input.");
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java:195]: Add more context to exception thrown

File: 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java [hunk 7]
```diff
@@ -155,6 +177,51 @@ private ConnectorResponseDTO createSecretManagerConnector(
...
+          ((CustomSecretManagerConnectorDTO) connectorConfigDTO).getTemplate().getVersionLabel(), false));
+      if (template == null || StringUtils.isBlank(template)) {
+        return;
+      }
+      JsonNode templateInputs = YamlUtils.read(template, JsonNode.class);
+      ArrayNode variablesArray = (ArrayNode) templateInputs.get("environmentVariables");
+      if (!EmptyPredicate.isEmpty(variablesArray)) {
+        for (JsonNode node : variablesArray) {
+          String key = node.get("name").asText();
+          // validate this key value is present in the incoming connector inputs
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java:210]: Did not understand this logic. 

Also lets divide this method in separate methods. Right now its difficult to understand.

File: 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java [hunk 8]
```diff
@@ -155,6 +177,51 @@ private ConnectorResponseDTO createSecretManagerConnector(
...
+          if (!inputs.contains(key)) {
+            log.error("Run time inputs of templates should be passed in connector.");
+            throw new InvalidRequestException("Run time inputs of templates should be passed in connector.");
+          }
+        }
+      }
+    }
+  }
+
+  private String templateIdFromRef(String templateRef) {
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java:220]: Please add UTs for all new methods

**Review by meenakshi-harness — CHANGES_REQUESTED** [review]

File: 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java (hunk 8 above)
Comment (VikasMaddukuriHarness) [reply 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java:228]: done

**Comment by VikasMaddukuriHarness** [comment]
trigger uts

**Comment by VikasMaddukuriHarness** [comment]
trigger messagecheck

**Comment by VikasMaddukuriHarness** [comment]
trigger messagecheck

**Review by meenakshi-harness — DISMISSED** [review]

File: 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java [hunk 9]
```diff
@@ -147,6 +162,14 @@ private ConnectorResponseDTO createSecretManagerConnector(
     // Check which type of token is provided
     ngVaultService.processTokenLookup(connector, accountIdentifier);
 
+    // Validate inputs for custom secret manager
+    try {
+      validateCustomSecretManagerInputs(connectorConfigDTO, accountIdentifier, connectorInfo.getOrgIdentifier(),
+          connectorInfo.getProjectIdentifier());
+    } catch (IOException ex) {
+      log.error("error reading templateInputs from template YAML which is used in this Custom Secret Manager");
```
[the remark this answers is not in the document]
Comment (VikasMaddukuriHarness) [reply 120-ng-manager/src/main/java/io/harness/ng/SecretManagerConnectorServiceImpl.java:170]: done

**Comment by VikasMaddukuriHarness** [comment]
trigger uts

**Comment by VikasMaddukuriHarness** [comment]
trigger uts

**Review by meenakshi-harness — APPROVED** [review]

**Pull request merged**